job/replay_syncer.go

```diff
@@ -0,0 +1,181 @@
+}
+
+func (s Syncer) Sync(context context.Context, runTimeout time.Duration) error {
+	replaySpecRepo := s.replaySpecFactory.New(models.JobSpec{})
+	replaySpecs, err := replaySpecRepo.GetByStatus(ReplayStatusToSynced)
+	if err != nil {
+		if err == store.ErrResourceNotFound {
+			return nil
+		}
+		return err
+	}
+
+	for _, replaySpec := range replaySpecs {
+		// sync end state of replayed replays
+		if replaySpec.Status == models.ReplayStatusReplayed {
+			if err := s.syncRunningReplay(context, replaySpec, replaySpecRepo); err != nil {
+				return err
+			}
+			continue
+		}
+
+		// sync timed out replays for accepted and in progress replays
+		if err := syncTimedOutReplay(replaySpecRepo, replaySpec, runTimeout); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func syncTimedOutReplay(replaySpecRepo store.ReplaySpecRepository, replaySpec models.ReplaySpec, runTimeout time.Duration) error {
+	runningTime := time.Now().Sub(replaySpec.CreatedAt)
+	if runningTime > runTimeout {
+		if updateStatusErr := replaySpecRepo.UpdateStatus(replaySpec.ID, models.ReplayStatusFailed, models.ReplayMessage{
+			Type:    ReplayRunTimeout,
+			Message: fmt.Sprintf("replay has been running since %s", replaySpec.CreatedAt.UTC().Format(TimestampLogFormat)),
+		}); updateStatusErr != nil {
+			logger.I(fmt.Sprintf("marking long running replay jobs as failed: %s", updateStatusErr))
+		}
+	}
+	return nil
+}
+
+func (s Syncer) syncRunningReplay(context context.Context, replaySpec models.ReplaySpec, replaySpecRepo store.ReplaySpecRepository) error {
+	var projSpec models.ProjectSpec
+	for _, project := range s.projects {
+		if project.Name == replaySpec.Job.Project.Name {
+			projSpec = project
+			break
+		}
+	}
+	jobSpecMap, err := s.prepareJobSpecMap(projSpec)
+	if err != nil {
+		return err
+	}
+
+	replayRequest := &models.ReplayRequest{
+		ID:         replaySpec.ID,
+		Job:        replaySpec.Job,
+		Start:      replaySpec.StartDate,
+		End:        replaySpec.EndDate,
+		JobSpecMap: jobSpecMap,
+	}
+
+	//for each replay, will get its list of jobs and runs and check if completed
+	treeNode, err := prepareReplayExecutionTree(replayRequest)
+	if err != nil {
+		return err
+	}
+	allNodes := treeNode.GetAllNodes()
+
+	stateSummary, err := s.checkInstanceState(context, projSpec, allNodes, replayRequest.Start, replayRequest.End)
+	if err != nil {
+		return err
+	}
+
+	return updateCompletedReplays(stateSummary, replaySpecRepo, replayRequest.ID)
+}
+
+func (s Syncer) prepareJobSpecMap(projSpec models.ProjectSpec) (map[string]models.JobSpec, error) {
+	projectJobSpecRepo := s.projectJobSpecRepoFactory.New(projSpec)
+	jobSpecs, err := projectJobSpecRepo.GetAll()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to retrieve jobs")
+	}
+
+	for i, jSpec := range jobSpecs {
+		if jobSpecs[i].Assets, err = s.assetCompiler(jSpec, s.Now()); err != nil {
+			return nil, errors.Wrap(err, "asset compilation")
+		}
+	}
+
+	var resolvedSpecs []models.JobSpec
```

Comment:
you are not really using this variable anywhere

Reply:
fixed